internals/overlord/logstate/gatherer.go

```diff
@@ -189,30 +178,51 @@ func (g *logGatherer) PlanChanged(pl *plan.Plan, serviceData map[string]*Service
 			continue
 		}
 
-		g.pullers.Add(service.Name, svcData.Buffer, g.entryCh)
-		// update labels
+		labels := evaluateLabels(g.target.Labels, serviceData[service.Name].Env)
```

Comment:
We've just been told the plan has changed, and yet we're evaluating this with labels cached from the prior plan. This looks like a bug?  The design we had before this change seemed to intentfully not cache unnecessary data, precisely so we could avoid this kind of pitfall by just not having the data at hand in the first place.

As a more general point, the recommendation on the prior review was to do less, by simply using the actual events we already had with the new parameter which included the environment variable.  So this PR is going quite a bit beyond it by introducing two new event methods here, and caching further data from the plan.  As observed, this creates further chances for bugs and discussion and moves us farther from the merging.

Can we backtrack a bit and just do the suggestions on the prior state so we can get this in?

Reply:
So there's a difference between:
1) the service's `environment` field as defined in the plan
2) the actual environment that the service is run inside

(2) is (1) with a bunch of defaults and extra variables added - see [serviceData.startInternal](https://github.com/canonical/pebble/blob/b255ff384ae0d8859706872e934e5f0b1ba750eb/internals/overlord/servstate/handlers.go#L350).

We were trying to use (2) instead of (1) for completeness. If you'd prefer we just use the environment as defined in the plan (and ignore all the default variables set by the service manager), I'm happy to do that instead. It certainly would simplify things.

Reply:
Jordan's updated this back to the simpler version that just uses the new plan.

Later, when we add [support for `$INTERPOLATION` in environment variables](https://github.com/canonical/pebble/issues/182) we can tweak this flow if needed.